### DISCUSSION TITLE
Warning for definite StackOverflowException

### DISCUSSION THREAD

**Discussion by michaelcheers** [dotnet/csharplang#5000] [Language Ideas]
Sometimes I am writing an overload for a function and I actually reference the function itself. It would be nice for it to give me a warning.

Something like:

```csharp
public void AddFile (Stream stream) => AddStream(stream.ToMemoryStream());
public void AddFile (MemoryStream stream) => AddStream(stream.ToMemoryStream()); // meant to do .ToArray()
public void AddFile (byte[] stream)
{
    // ...
}
```

I propose that it warn me that the second function calls itself.

**Comment by HaloFour** [comment]
Sounds like a good use case for an analyzer.

**Reply by CyrusNajmabadi** [reply]
note: we've looked into this in the past, and there were a lot of false positives.  specifically, because work done did have side-effects that made it ok for the method to call back into itself.   The most reasonable case to warn on is the non-virtual case that directly calls itself passing along the args it had exactly with no other logic.

**Reply by HaloFour** [reply]
This is what IntelliJ offers:

![image](https://user-images.githubusercontent.com/10102290/128250265-7b5a6b87-1c50-4faa-9ccf-24cdf33e852b.png)

The glyph in the gutter indicates any recursive call, which would probably go a long way to help the developer identify the case.

**Reply by michaelcheers** [reply]
I think it should be only for non-virtual members and only if all code path run themselves.

**Comment by michaelcheers** [comment]
Just found another use case for this.

Say I have a function like this:
```csharp
public static MemberInstantiation FindSymbol (ISymbol sym)
    => allSymbols[sym.ContainingType!].Children.Single(mi => mi.Symbol == sym);
```
It is easy to make mistakes while writing a casting function. Eg.
```csharp
public static MethodInstantiation FindSymbol (IMethodSymbol sym)
    => (MethodInstantiation)FindSymbol(sym);
```
This code looks almost fine and compiles with no problems but it will give a `StackOverflowException` if actually used.

Often, I will make this mistake and I have to change it to:
```csharp
public static MethodInstantiation FindSymbol (IMethodSymbol sym)
    => (MethodInstantiation)FindSymbol(sym);
```

This along with [not stepping into auto-properties in the debugger](https://developercommunity.visualstudio.com/t/Step-into-properties-option-should-skip/578674?space=8&q=auto+properties+debugger) would save me a lot of time.
